MathlibTest/Junk.lean

```diff
@@ -0,0 +1,360 @@
+`fun x y ↦ exp (y * log x)`. The resulting definition for reals can be expressed in terms of
+`Real.exp`, `Real.log`, and `Real.cos`. In particular, this does _not_ agree with the schoolbook
+definition of odd roots of negative numbers, i.e. `(-1) ^ (1/3) ≠ -1`.
+
+In this example, `(-8) ^ (5/3) = 16`. Contrast with how it might "conventionally" have the value of
+`((-8) ^ (1/3)) ^ 5 = (-2) ^ 5 = -32`.
+-/
+example : (-8 : ℝ) ^ (5/3 : ℝ) = 16 := by
+  rw [Real.rpow_def_of_neg (by norm_num), Real.log_neg_eq_log]
+  rw [← Real.cos_two_pi_sub, ← sub_mul, mul_comm (_ - _)]
+  norm_num
+  rw [mul_div Real.pi, mul_one, Real.cos_pi_div_three]
+  suffices Real.exp (Real.log 8 * (5 / 3)) = Real.exp (Real.log 32) by
+    rw [this, Real.exp_log (by positivity)]
+    norm_num
+  rw [show (8 : ℝ) = 2 ^ 3 by norm_num, show (32 : ℝ) = 2 ^ 5 by norm_num]
+  simp only [Real.log_pow, Nat.cast_ofNat]
+  ring_nf
+
+/-- The binary entropy function `Real.binEntropy` is simply defined in terms of `Real.log` as
+`p * log p⁻¹ + (1 - p) * log (1 - p)⁻¹`, so it happily provides junk values outside of its
+traditional domain `[0, 1]`. In this example, the entropy of the "probability" `-1` turns out to be
+`-2 * log 2`. -/
+example : Real.binEntropy (-1) = -2 * Real.log 2 := by
+  simp only [Real.binEntropy, Real.log_inv]
+  norm_num
+
+/-! ## Arcsin and arccos -/
+
+/-- Several functions on reals are defined using `OrderIso.symm`, for instead, `Real.arcsin`. This
+means that the definition proceeds by starting with `Real.sin`, observing (in `Real.sinOrderIso`)
+that it's a strictly monotonic and continuous map from `[-π/2, π/2]` to `[-1, 1]`, thus, a
+bijection. Then inverting this function gives a map from `[-1, 1]` to `[-π/2, π/2]`. Finally, this is
+_extended_ to the whole reals, taking the extreme points `-π/2` and `π/2` as the minimal extension.
+
+This means that `Real.arcsin x = -π/2` for all `x ≤ -1`, and it's `π/2` for all `x ≥ 1`.
+-/
+example : Real.arcsin (-4) = -(Real.pi/2) ∧ Real.arcsin 100 = Real.pi/2 := by
+  norm_num
+
+/-- Similarly, `Real.arccos` has the junk values of `π` on arguments less below `-1`, and `0` on
+arguments above `1`. -/
+example : Real.arccos (-37) = Real.pi ∧ Real.arccos 1037 = 0 := by
+  norm_num
+
+/-! ## Indexing and lists -/
+
+/-- Accessing a list `l : List α` with an "out of bounds" index is possible using the syntax `l[i]?` or
+`l[i]!`. These are notations for `GetElem?.getElem?` and `GetElem?.getElem!`, respectively. The
+former is "safe": it returns an `Option α`, which is `Option.none` if the index was out of bounds.
+
+But `getElem!` takes an `Inhabited` instance on the type `α` and uses that as the junk value for
+any extra indexing. In this example, we make a list of two natural numbers `[3, 5]` and access index
+100. This gives the `Inhabited.default` value for `Nat`, which happens to be 0.
+-/
+example : [3, 5][100]! = default ∧ default = 0 := by
```

Comment:
this is very implicit that the RHS of the And is an equality of Nats. Maybe a type ascription on the default is called for.